### PR TITLE
AME-194: Add Advanced Matching Parameters and template tests

### PR DESCRIPTION
**Pull request by monicating**
This PR:
- Adds Advanced Matching parameters to the template
  - Advanced parameters currently include: `email`, `aaid`, `idfa`
  - These parameters need to be pushed to the dataLayer, for example:
  ```      
      window.dataLayer.push({
        email: 'alice@example.com',
        aaid: 'cdda802e-fb9c-47ad-9866-0794d394c912',
        idfa: 'EA7583CD-A667-48BC-B806-42ECB2B48606'      
      });
  ```

- Adds tests for the template

Updated Reddit Tag creation screenshot:
pixel-screenshot

### CODE REVIEW

File: template.tpl [hunk 1]
```diff
@@ -436,7 +505,165 @@ ___WEB_PERMISSIONS___
...
+    runCode(mockData);
+
+    // Verify that the tag finished successfully.
+    assertApi('gtmOnSuccess').wasCalled();
+- name: Test Tracking Event Type
+  code: "mock('copyFromWindow', key => {\n  if (key === 'rdt') return function() {\n\
+    \    if (arguments[0] === 'track') {\n      assertThat(arguments[1], 'Incorrect\
+    \ Tracking Event Type').isEqualTo(mockData.eventType);\n    }\n  };\n});\n   \
+    \  \n// Call runCode to run the template's code.\nrunCode(mockData);\n\n// Verify\
+    \ that the tag finished successfully.\nassertApi('gtmOnSuccess').wasCalled();"
```
Comment (dmaii) [inline template.tpl:652]: Any idea why this is so weirdly spaced?

File: template.tpl [hunk 2]
```diff
@@ -102,6 +102,72 @@ ___TEMPLATE_PARAMETERS___
...
+      }
+    ],
+    "enablingConditions": [
+      {
+        "paramName": "advancedMatching",
+        "paramValue": true,
+        "type": "EQUALS"
+      }
+    ],
+    "help": "Advanced Matching Parameters must be pushed to the Data Layer."
```
Comment (dmaii) [inline template.tpl:170]: Did marketing approve this help text? I think we may want to make a follow-up ticket to also update https://ads.reddit.com/account/11qnzrqv/conversions#gtm with some additional instructions.

File: template.tpl [hunk 3]
```diff
@@ -436,7 +505,165 @@ ___WEB_PERMISSIONS___
...
+    assertApi('gtmOnSuccess').wasCalled();
+- name: Test Advanced Matching
+  code: |
+    mockData = {
+      id: "t2_123",
+      event_type: "PageVisit",
+      enableFirstPartyCookies: true,
+      advancedMatching: true,
+      advancedMatchingParams: [
+        {name: 'optOut', value: 1},
```
Comment (dmaii) [inline template.tpl:605]: I believe this is actually a boolean field that's converted to a 1 or 0?

https://github.snooguts.net/reddit/reddit-conversion-pixel-js/blob/develop/src/lib/pixelLib.js#L93

**Review by dmaii — CHANGES_REQUESTED** [review]

**Comment by dmaii** [comment]
Thanks for adding all these tests!

**Comment by dmaii** [comment]
🔕 If there's some way to validate the actual requests being made, that would be phenomenal.

File: template.tpl (hunk 3 above)
Comment (woodchuckgym) [reply template.tpl:605]: The value of optOut can be any [truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) or [falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) value.  If optOut is not specified, it defaults to false (and permits retargeting).

File: template.tpl [hunk 4]
```diff
@@ -133,9 +195,11 @@ var getRdt = function() {
   return copyFromWindow('rdt');
 };
 
+var advancedParams = data.advancedMatchingParams && data.advancedMatchingParams.length ? makeTableMap(data.advancedMatchingParams, 'name', 'value') : {};
+
 var _rdt = getRdt();
 if (!_rdt.advertiserId) {
-  _rdt('init', data.id);
+  _rdt('init', data.id, advancedParams);
```
Comment (woodchuckgym) [inline template.tpl:202]: Can you add `integration: 'gtm'` to the initialization parameters?

**Review by woodchuckgym — CHANGES_REQUESTED** [review]

File: template.tpl (hunk 4 above)
Comment (monicating) [reply template.tpl:202]: Ah yeah, I had this change ready in my workspace but thought I would hold off until the pixel library changes were released. Although adding it here now or not would not really make a difference.

Comment (monicating) [reply template.tpl:202]: I forgot I had this PR open: https://github.com/monicating/reddit-gtm-template/pull/1/files

**Review by dmaii — APPROVED** [review]

File: template.tpl (hunk 4 above)
Comment (woodchuckgym) [reply template.tpl:202]: Ah, ok.  Cool!  Sounds good to me.  I didn't know if a merge here kicked off an automatic update to teh Google.  TY!

**Review by woodchuckgym — APPROVED** [review]

**Comment by monicating** [comment]
Closing in favor of: https://github.com/reddit/reddit-gtm-template/pull/5

**Pull request closed without merging**